systemd_ctypes/bus.py

```diff
@@ -320,124 +327,467 @@ def add_match(self, rule, handler):
-            handled = self.handle_method_call(message, reply)
-        except BusError:
-            assert False  # needs proper exception handling
-
-        if handled:
-            # message_send() is not available in RHEL 8
-            message.get_bus().send(reply, None)
-            return 1
-        else:
-            return 0
+    """Base object type for exporting objects on the bus
+
+    This is the lowest-level class that can be passed to Bus.add_object().
+
+    If you want to directly subclass this, you'll need to implement
+    `message_received()`.
+
+    Subclassing from `bus.Object` is probably a better choice.
+    """
+    __dbus_bus: Optional[Bus] = None
+    __dbus_path: Optional[str] = None
+
+    def registered_on_bus(self, bus: Bus, path: str) -> None:
+        """Report that an instance was exported on a given bus and path.
+
+        This is used so that the instance knows where to send signals.
+        Bus.add_object() calls this: you probably shouldn't call this on your
+        own.
+        """
+        self.__dbus_bus = bus
+        self.__dbus_path = path
+
+        self.registered()
+
+    def registered(self) -> None:
+        """Called after an object has been registered on the bus
+
+        This is the correct method to implement to do some initial work that
+        needs to be done after registration.  The default implementation does
+        nothing.
+        """
+        pass
+
+    def emit_signal(self, interface: str, name: str, signature: str, *args: Any) -> bool:
+        """Emit a D-Bus signal on this object
+
+        The object must have been exported on the bus with Bus.add_object().
+
+        :interface: the interface of the signal
+        :name: the 'member' name of the signal to emit
+        :signature: the type signature, as a string
+        :args: the arguments, according to the signature
+        :returns: True
+        """
+        assert self.__dbus_bus is not None and self.__dbus_path is not None
+        return self.__dbus_bus.message_new_signal(self.__dbus_path, interface, name, signature, *args).send()
+
+    def message_received(self, message: BusMessage) -> bool:
+        """Called when a message is received for this object
+
+        This is the lowest level interface to the BaseObject.  You need to
+        handle method calls, properties, and introspection.
 
-    def handle_method_call(self, message, reply):
+        You are expected to handle the message and return True.  Normally this
+        means that you send a reply.  If you don't want to handle the message,
+        return False and other handlers will have a chance to run.  If no
+        handler handles the message, systemd will generate a suitable error
+        message and send that, instead.
+
+        :message: the message that was received
+        :returns: True if the message was handled
+        """
         raise NotImplementedError
 
-class Object(BaseObject):
+
+class Interface:
```

Comment:
I don't immediately see the point of the `Interface` class.  `Object` inherits from it anyway.  Can't we just move its content into `Object` and simplify the API?

Reply:
The issue is that we need the `__init_subclass__` below the various interfaces that we define, but `Object` needs to contain those interface.